README.md
# MMM-LocalTemperature

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/) smart mirror project.

This module reads and displays temperature and humidity information from a sensor connected to a Raspberry Pi GPIO pin.  

| Status  | Version | Date       | Maintained? | Minimum MagicMirror² Version |
|:------- |:------- |:---------- |:----------- |:---------------------------- |
| Working | `1.4.0` | 2021-09-19 | Yes         |`2.2.1`                       |

### Example
![Example of MMM-LocalTemperature](images/sample.png?raw=true "Example screenshot")

### Notable Features
1. Get data from a DHT11, DHT22, or AM2302 sensor
2. Display the temperature and/or humidity from this module and/or,
3. Send the temperature to the built in 'weather' module via module notifications

### Dependencies
1. A local 'c' program, used to read the data from the sensor (included)
2. The WiringPi library (Free to install and use)

## Installation
To install the module, use your terminal to:
1. Navigate to your MagicMirror's modules folder. If you are using the default installation directory, use the command:<br />`cd ~/MagicMirror/modules`
2. Copy the module to your computer by executing the following command:<br />`git clone https://github.com/glitch452/MMM-LocalTemperature.git`
3. Install the WiringPi library by executing the following command:<br />`sudo apt-get -y update && sudo apt-get -y upgrade && sudo apt-get -y install build-essential wiringpi`
    **Note:** If running on Raspberry Pi 4, use the following process to update wiringpi  
    [http://wiringpi.com/wiringpi-updated-to-2-52-for-the-raspberry-pi-4b](http://wiringpi.com/wiringpi-updated-to-2-52-for-the-raspberry-pi-4b)
4. Make sure the `DHT` program that reads the sensor data is executable by executing the following command:<br />`cd MMM-LocalTemperature && chmod +x DHT`

## Using the module

### MagicMirror² Configuration

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        ...
        {
            module: "MMM-LocalTemperature",
            position: "top_left", // Only add a position if you want this module to display the data
            header: "Room Temperature",
            config: {
                sensorPin: 22, // For GPIO 22
                ...
                // See below for more Configuration Options
            }
        },
        ...
    ]
}
```

### Configuration Options

| Option                  | Details
|:----------------------- |:-------------
| `sensorPin`             | **REQUIRED** - The GPIO Pin number that is connected to the data pin on the sensor. The default pin scheme is the standard Raspberry Pi (BCM) GPIO numbering system for Rev 2 Pi's. See the `pinScheme` option for other numbering systems.<br />**Type:** `number`
| `pinScheme`             | *Optional* - The pin numbering system to use for the `sensorPin` option. See this [interactive pinout diagram](https://pinout.xyz) for more details on pin usage for the Raspberry Pi. <br />Note: Ultimately the `sensorPin` value will be converted to the WiringPi system, becuase that is the library used by the `DHT` program to interact with the pin. However, any of these numbering systems can be used, since this module will convert the `sensorPin` value automatically based on the selected scheme. <br />**Type:** `string`<br />**Default:** `"BCMv2"`<br />**Options:**<br />- `"BCMv2"` The standard Raspberry Pi GPIO numbering system on current (Rev 2) boards<br />- `"BCMv1"` The standard Raspberry Pi GPIO numbering system on older (Rev 1) boards<br />- `"BOARD"` The physical pin numbering on the GPIO header<br />- `"WPI"` The WiringPi numbering system
| `useSudo`               | *Optional* - Whether or not to use sudo when calling the script to get the sensor data. A value of `true` uses sudo, `false` does not use sudo.  (Added in v1.4.0)<br />**Type:** `boolean`<br />**Default:** `false`
| `units`                 | *Optional* - The unit system to use for the temperature value. (`"metric"` = Celcius, `"imperial"` = Fahrenheit, `"default"` = Kelvin)<br />**Type:** `string`<br />**Default:** `config.units`<br />**Options:** `"metric"`, `"imperial"`, `"default"`
| `sendTemperature`       | *Optional* - When `true`, an "INDOOR_TEMPERATURE" notification is sent to the other modules when the data is received from the sensor.  This can be used to display the indoor temperature within the built-in 'currentweather' module. The 'currentweather' module's `showIndoorTemperature` option must be set to `true` for it to display the data sent from this module.<br />**Type:** `boolean`<br />**Default:** `true`
| `sendHumidity`          | *Optional* - When `true`, an "INDOOR_HUMIDITY" notification is sent to the other modules when the data is received from the sensor.  This can be used to display the indoor humidity within the built-in 'currentweather' module. The 'currentweather' module's `showIndoorHumidity` option must be set to `true` for it to display the data sent from this module.<br />**Type:** `boolean`<br />**Default:** `true`
| `showTemperature`       | *Optional* - When `true`, the module will display the temperature on screen.<br />**Type:** `boolean`<br />**Default:** `false`
| `showHumidity`          | *Optional* - When `true`, the module will display the humidity on screen.<br />**Type:** `boolean`<br />**Default:** `false`
| `temperatureOffset`     | *Optional* - A value to add to the temperature value from the sensor (in the selected units).  It can be positive or negative. This can be used to adjust for a sesor that is consistently off by a specific amount.<br />**Type:** `number`<br />**Default:** `0`
| `humidityOffset`        | *Optional* - A value to add to the humidity value from the sensor. It can be positive or negative. This can be used to adjust for a sesor that is consistently off by a specific amount.<br />**Type:** `number`<br />**Default:** `0`
| `iconView`              | *Optional* - When `true`, a view which uses icons and the data will be shown instead of the standard temperature and humidity text. The data shown depends on the `showTemperature` and `showHumidity` options. <br />**Type:** `boolean`<br />**Default:** `true`
| `temperatureText`       | *Optional* - The text template to be used when displaying the temperature data. The stings `"{temperature}"` and `"{humidity}"` will be replaced with the temperature and humidity values respectively. For icons, `"{icon-<fa_name>}"` will be replaced with the html tag for the corresponding [FontAwesome](https://fontawesome.com/icons?d=gallery&m=free) icon (this uses the solid style).  Use `"{icon-regular-<fa_name>}"`, `"{icon-solid-<fa_name>}"`, `"{icon-brand-<fa_name>}"` to specify the solid / regular type or when using a brand icon.  Ex: use `"{icon-solid-thermometer-half}"` for the fa-thermometer-half icon.  <br />**Type:** `string`<br />**Default:** `"Temperature: {temperature}°C/°F/K"`
| `humidityText`          | *Optional* - The text template to be used when displaying the humidity data. The stings `"{temperature}"` and `"{humidity}"` will be replaced with the temperature and humidity values respectively. For icons, `"{icon-<fa_name>}"` will be replaced with the html tag for the corresponding [FontAwesome](https://fontawesome.com/icons?d=gallery&m=free) icon (this uses the solid style).  Use `"{icon-regular-<fa_name>}"`, `"{icon-solid-<fa_name>}"`, `"{icon-brand-<fa_name>}"` to specify the solid / regular type or when using a brand icon.  Ex: use `"{icon-solid-thermometer-half}"` for the fa-thermometer-half icon.  <br />**Type:** `string`<br />**Default:** `"Humidity: {humidity}%"`
| `fontSize`              | *Optional* - The main font size to use for the module text. <br />**Type:** `string`<br />**Default:** `'medium'`<br />**Options:** `'x-small'`, `'small'`, `'medium'`, `'large'`, `'x-large'`
| `decimalSymbol`         | *Optional* - The character to use as the decimal symbol. <br />**Type:** `string`<br />**Default:** `"."`
| `roundTemperature`      | *Optional* - When true, the temperature value will be rounded to the nearest integer. <br />**Type:** `boolean`<br />**Default:** `false`
| `roundHumidity`         | *Optional* - When true, the humidity value will be rounded to the nearest integer. <br />**Type:** `boolean`<br />**Default:** `false`
| `scriptPath`            | *Optional* - The location of the `DHT` program. This is the fully qualified path including the program file name. <br />**Type:** `string`<br />**Default:** `<module_folder_path>\DHT`
| `updateInterval`        | *Optional* - The number of minutes to wait before requesting an update of the data from the sensor. The minimum value is `0.5`. <br />**Type:** `number`<br />**Default:** `5`
| `retryDelay`            | *Optional* - The number of seconds to wait before trying to request the sensor data after a sensor read failure. The minimum value is `10`. Since the reading of the data from the sensor is time sensitive, and can be interrupted by context switching, the `DHT` program is told to try up to 3 times.  This option is the amount of time for this module to wait before the next 3 tries. <br />**Type:** `number`<br />**Default:** `10`
| `initialLoadDelay`      | *Optional* - The number of seconds to wait before starting to run this module. The minimum value is `0`. <br />**Type:** `number`<br />**Default:** `0`
| `animationSpeed`        | *Optional* - The number of milliseconds to use for the animation when updating the on-screen display of this module. The minimum value is `0`.<br />**Type:** `number`<br />**Default:** `0`

## Connecting the Sensor to the Raspberry Pi

Here are some diagrams that you may find useful when connecting your sensor.  See the [guide from adafruit.com](docs/adafruit-dht-sensor-guide.pdf?raw=true) for more wiring and programming information.

### DHT11 sensor
![Example of a DHT11 sensor with a Raspberry Pi](images/wiring_dht11.png?raw=true "How to connect a DHT11 sensor to a Raspberry Pi")

### DHT22 / AM2302 sensor
![Example of a DHT22 / AM2302 sensor with a Raspberry Pi](images/wiring_dht22.png?raw=true "How to connect a DHT22 sensor to a Raspberry Pi")

### AM2302 sensor with wires
![Example of a AM2302 sensor with a Raspberry Pi](images/wiring_am2302.png?raw=true "How to connect a AM2302 sensor to a Raspberry Pi")

## Customizing the DHT Program

The `DHT` program is a `c` coded program used to read the sensor data. A `c` program is required to read the sensor data because the data is transmitted by extremely short pulses of signal and the `c` code executes much faster than python.

The included `DHT` program is something I pieced together based on code from [dht22 by nebulx29](https://github.com/nebulx29/dht22) and the [Python DHT Sensor Library by Adafruit](https://github.com/adafruit/Adafruit_Python_DHT).

If you would like to customize the `DHT` program, you can edit the file `DHT.c` with your favorite editor, then re-compile it using the following command: `cc -Wall DHT.c -o DHT -lwiringPi`

## Updates
To update the module to the latest version, use your terminal to:
1. Navigate to your MMM-LocalTemperature folder. If you are using the default installation directory, use the command:<br />`cd ~/MagicMirror/modules/MMM-LocalTemperature`
2. Update the module by executing the following command:<br />`git pull`

If you have changed the module on your own, the update will fail. <br />To force an update (WARNING! your changes will be lost), reset the module and then update with the following commands:
```
git reset --hard
git pull
```

## Manually Choose a Version

To use an older version of this module, use your terminal to:
1. Navigate to your MMM-LocalTemperature folder. If you are using the default installation directory, use the command:<br />`cd ~/MagicMirror/modules/MMM-LocalTemperature`
2. Fetch all the available tags<br />`git fetch`
3. Show all the available tags<br />`git tag`
4. Checkout one of the available tags<br />`git checkout {tag_name}`<br />Example: `git checkout v1.0.0`


To switch back to the latest version, use your terminal to:
1. Navigate to your MMM-LocalTemperature folder. If you are using the default installation directory, use the command:<br />`cd ~/MagicMirror/modules/MMM-LocalTemperature`
2. Checkout the master branch<br />`git checkout master`

## License

### The MIT License (MIT)

Copyright © 2018 David Dearden

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the “Software”), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

**The software is provided “as is”, without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and noninfringement. In no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software.**

CHANGELOG.md
# MMM-LocalTemperature Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.4.0] - 2021-09-19

### Added
- An option to decide whether or not to use `sudo` when calling the sensor script
- Spanish Translation (thanks to [rafagale](https://github.com/rafagale))
- The ability to set a temperature and humidity offset

### Fixed
- The DOM may try to load when there is no sensor data

## [1.3.0] - 2019-11-13

### Added
- Added the ability to specify icons in the `temperatureText` and `humidityText` options

### Changed
- Updated the readme to explain how to use the new icon feature

### Fixed
- Bug fixed for the animation speed value validation

## [1.2.0] - 2019-05-22

### Added
- French Translation (thanks to [laventin85](https://github.com/laventin85))
- Number validation now accounts for NaN values
- README instructions for choosing a specific version of this module

### Fixed
- Console error when resuming the module (lastUpdateTime variable initialized and updated)
- Check for 'object' type when checking for axis.js script
- Hide "Socket Notification Received" message for non-development mode

## [1.1.0] - 2018-06-05

### Added
- Icon View

## [1.0.0] - 2018-06-04

Initial Release of the MMM-LocalTemperature module. 

## [0.2.0] - 2018-05-27

Pre-Release of the MMM-LocalTemperature module for testing purposes. 

## [0.1.0] - 2018-05-27

Pre-Release of the MMM-LocalTemperature module for testing purposes. 


MMM-LocalTemperature.js
/* global Module */

/**
 * Magic Mirror
 * Module: MMM-LocalTemperature
 *
 * By David Dearden
 * MIT Licensed.
 */

var axis, Log, config;

/**
 * Register the module with the MagicMirror program
 */
Module.register("MMM-LocalTemperature", {

	/**
	 * The default configuration options
	 */
	defaults: {
		sensorPin: null,
		pinScheme: "BCMv2",
		units: config.units,
		useSudo: false,
		sendTemperature: true,
		sendHumidity: true,
		showTemperature: false,
		showHumidity: false,
		temperatureOffset: 0,
		humidityOffset: 0,
		iconView: true,
		temperatureText: null, // Set in self.start() becuase access to self.translate is needed
		humidityText: null, // Set in self.start() becuase access to self.translate is needed
		fontSize: "medium",
		decimalSymbol: null, // Set in self.start() becuase access to self.translate is needed
		roundTemperature: false,
		roundHumidity: false,
		scriptPath: null, // Set in self.start() becuase access to self.data.path is needed
		initialLoadDelay: 0, // Seconds, minimum 0
		animationSpeed: 0, // Milliseconds, minimum 0
		retryDelay: 10, // Seconds, minimum 10
		updateInterval: 5, // Minutes, minimum 0.5
		developerMode: false
	},

	/**
	 * The minimum version of magic mirror that is required for this module to run.
	 */
	requiresVersion: "2.2.1",

	/**
	 * Override the start function.  Set some instance variables and validate the selected
	 * configuration options before loading the rest of the module.
	 */
	start: function() {
		var self = this;
		self.instanceID = self.identifier + "_" + Math.random().toString().substring(2);
		self.sendSocketNotification("INIT", { instanceID: self.instanceID } );
		self.updateTimer = null;
		self.sensorData = null;
		self.loaded = false;
		self.defaults.scriptPath = self.data.path + "DHT";
		self.defaults.decimalSymbol = self.translate("DECIMAL_SYMBOL");
		self.maxDataAttempts = 3;
		self.validUnits = [ "metric", "imperial", "default" ];
		var unitMap = { "metric": "celcius", "imperial": "fahrenheit", "default": "kelvin" };
		self.validPinSchemes = [ "BOARD", "BCMv1", "BCMv2", "WPI" ];
		self.validFontSizes = [ "x-small", "small", "medium", "large", "x-large" ];
		self.currentweatherLoaded = false;
		self.lastUpdateTime = new Date(0);

		var pinMapping = [
			{ "BOARD": 10, "BCMv1": 15, "BCMv2": 15, "WPI": 16 },
			{ "BOARD": 11, "BCMv1": 17, "BCMv2": 17, "WPI": 0 },
			{ "BOARD": 12, "BCMv1": 18, "BCMv2": 18, "WPI": 1 },
			{ "BOARD": 13, "BCMv1": 21, "BCMv2": 27, "WPI": 2 },
			{ "BOARD": 15, "BCMv1": 22, "BCMv2": 22, "WPI": 3 },
			{ "BOARD": 16, "BCMv1": 23, "BCMv2": 23, "WPI": 4 },
			{ "BOARD": 18, "BCMv1": 24, "BCMv2": 24, "WPI": 5 },
			{ "BOARD": 19, "BCMv1": 10, "BCMv2": 10, "WPI": 12 },
			{ "BOARD": 21, "BCMv1": 9, "BCMv2": 9, "WPI": 13 },
			{ "BOARD": 22, "BCMv1": 25, "BCMv2": 25, "WPI": 6 },
			{ "BOARD": 23, "BCMv1": 11, "BCMv2": 11, "WPI": 14 },
			{ "BOARD": 24, "BCMv1": 8, "BCMv2": 8, "WPI": 10 },
			{ "BOARD": 26, "BCMv1": 7, "BCMv2": 7, "WPI": 11 },
			{ "BOARD": 29, "BCMv1": null, "BCMv2": 5, "WPI": 21 },
			{ "BOARD": 3, "BCMv1": 0, "BCMv2": 2, "WPI": 8 },
			{ "BOARD": 31, "BCMv1": null, "BCMv2": 6, "WPI": 22 },
			{ "BOARD": 32, "BCMv1": null, "BCMv2": 12, "WPI": 26 },
			{ "BOARD": 33, "BCMv1": null, "BCMv2": 13, "WPI": 23 },
			{ "BOARD": 35, "BCMv1": null, "BCMv2": 19, "WPI": 24 },
			{ "BOARD": 36, "BCMv1": null, "BCMv2": 16, "WPI": 27 },
			{ "BOARD": 37, "BCMv1": null, "BCMv2": 26, "WPI": 25 },
			{ "BOARD": 38, "BCMv1": null, "BCMv2": 20, "WPI": 28 },
			{ "BOARD": 40, "BCMv1": null, "BCMv2": 21, "WPI": 29 },
			{ "BOARD": 5, "BCMv1": 1, "BCMv2": 3, "WPI": 9 },
			{ "BOARD": 7, "BCMv1": 4, "BCMv2": 4, "WPI": 7 },
			{ "BOARD": 8, "BCMv1": 14, "BCMv2": 14, "WPI": 15 }
		];

		// Process and validate configuration options
		if (axis.isNumber(self.config.updateInterval) && !isNaN(self.config.updateInterval) && self.config.updateInterval >= 0.5) { self.config.updateInterval = self.config.updateInterval * 60 * 1000; }
		else { self.config.updateInterval = self.defaults.updateInterval * 60 * 1000; }
		if (axis.isNumber(self.config.retryDelay) && !isNaN(self.config.retryDelay) && self.config.retryDelay >= 10) { self.config.retryDelay = self.config.retryDelay * 1000; }
		else { self.config.retryDelay = self.defaults.retryDelay * 1000; }
		if (axis.isNumber(self.config.initialLoadDelay) && !isNaN(self.config.initialLoadDelay) && self.config.initialLoadDelay >= 0) { self.config.initialLoadDelay = self.config.initialLoadDelay * 1000; }
		else { self.config.initialLoadDelay = self.defaults.initialLoadDelay * 1000; }
		if (!axis.isNumber(self.config.animationSpeed) || isNaN(self.config.animationSpeed) || self.config.animationSpeed < 0) { self.config.animationSpeed = self.defaults.animationSpeed; }
		if (!axis.isString(self.config.scriptPath) || self.config.scriptPath.length < 1 ) { self.config.scriptPath = self.defaults.scriptPath; }
		if (!self.validUnits.includes(self.config.units)) { self.config.units = self.defaults.units; }
		self.tempUnit = unitMap[self.config.units];
		if (self.tempUnit === "celcius") {
			self.defaults.temperatureText = self.translate("SHOW_TEMP_CELCIUS", { "temperature_var": "{temperature}" });
		} else if (self.tempUnit === "fahrenheit") {
			self.defaults.temperatureText = self.translate("SHOW_TEMP_FAHRENHEIT", { "temperature_var": "{temperature}" });
		} else {
			self.defaults.temperatureText = self.translate("SHOW_TEMP_KELVIN", { "temperature_var": "{temperature}" });
		}
		self.defaults.humidityText = self.translate("SHOW_HUMIDITY", { "humidity_var": "{humidity}" });
		if (!axis.isString(self.config.temperatureText) || self.config.temperatureText.length < 1 ) { self.config.temperatureText = self.defaults.temperatureText; }
		if (!axis.isString(self.config.humidityText) || self.config.humidityText.length < 1 ) { self.config.humidityText = self.defaults.humidityText; }
		if (!self.validPinSchemes.includes(self.config.pinScheme)) { self.config.pinScheme = self.defaults.pinScheme; }
		if (!axis.isNumber(self.config.sensorPin) || isNaN(self.config.sensorPin)) { self.config.sensorPin = self.defaults.sensorPin; }
		if (!axis.isBoolean(self.config.useSudo)) { self.config.useSudo = self.defaults.useSudo; }
		if (!axis.isBoolean(self.config.sendTemperature)) { self.config.sendTemperature = self.defaults.sendTemperature; }
		if (!axis.isBoolean(self.config.sendHumidity)) { self.config.sendHumidity = self.defaults.sendHumidity; }
		if (!axis.isBoolean(self.config.roundTemperature)) { self.config.roundTemperature = self.defaults.roundTemperature; }
		if (!axis.isBoolean(self.config.roundHumidity)) { self.config.roundHumidity = self.defaults.roundHumidity; }
		if (!axis.isBoolean(self.config.iconView)) { self.config.iconView = self.defaults.iconView; }
		if (!axis.isString(self.config.decimalSymbol)) { self.config.decimalSymbol = self.defaults.decimalSymbol; }
		if (!self.validFontSizes.includes(self.config.fontSize)) { self.config.fontSize = self.defaults.fontSize; }
		if (axis.isNumber(self.config.temperatureOffset) && !isNaN(self.config.temperatureOffset)) { self.config.temperatureOffset = self.config.temperatureOffset; }
		else { self.config.temperatureOffset = self.defaults.temperatureOffset; }
		if (axis.isNumber(self.config.humidityOffset) && !isNaN(self.config.humidityOffset)) { self.config.humidityOffset = self.config.humidityOffset; }
		else { self.config.humidityOffset = self.defaults.humidityOffset; }

		// Validate the provided sensorPin
		var pinObj = pinMapping.find(function(val) { return val[this.scheme] === this.pin; }, { scheme: self.config.pinScheme, pin: self.config.sensorPin });
		if (axis.isUndefined(pinObj)) {
			self.log(self.translate("INVALID_PIN", { "pinValue": self.config.sensorPin }), "error");
			self.config.sensorPin = null;
		} else {
			// Select the WiringPi pin number
			self.config.sensorPin = pinObj.WPI;
		}

		self.log(("start(): self.data: " + JSON.stringify(self.data)), "dev");
		self.log(("start(): self.config: " + JSON.stringify(self.config)), "dev");

		// Start this module - Request the data from the sensor
		if (!axis.isNull(self.config.sensorPin)) {
			if (self.config.initialLoadDelay > 0) {
				self.log(self.translate("INITIAL_DELAY", { "seconds": (self.config.initialLoadDelay / 1000) }));
				setTimeout(function(){ self.getData(1); self.scheduleUpdate(); }, self.config.initialLoadDelay );
			} else {
				self.getData(1);
				self.scheduleUpdate();
			}
		}

	},

	/**
	 * Override the suspend function that is called when the module instance is hidden.
	 * This method stops the update timer.
	 */
	suspend: function() {
		var self = this;
		self.log(self.translate("SUSPENDED") + ".");
		clearInterval(self.updateTimer);
	},

	/**
	 * Override the resume function that is called when the module instance is un-hidden.
	 * This method re-starts the update timer and calls for an update if the update interval
	 * has been passed since the module was suspended.
	 */
	resume: function() {
		var self = this;
		self.log(self.translate("RESUMED") + ".");
		self.scheduleUpdate();
		var date = new Date();
		var threshold = new Date( self.lastUpdateTime.getTime() + self.config.updateInterval );
		if (date >= threshold) { self.getData(1); }
	},

	/**
	 * The scheduleUpdate function starts the auto update timer.
	 */
	scheduleUpdate: function() {
		var self = this;
		self.updateTimer = setInterval(function() { self.getData(1); }, self.config.updateInterval);
		self.log( self.translate("UPDATE_SCHEDULED", { "minutes": (self.config.updateInterval / (1000 * 60)) }) );
	},

	/**
	 * The getData function sends a request to the node helper read the data from the sensor
	 *
	 * @param attemptNum (number) The number of attempts to read the sensor data
	 */
	getData: function(attemptNum) {
		var self = this;
		self.log(self.translate("DATA_REQUESTED"));
		self.sendSocketNotification("GET_DATA", {
			instanceID: self.instanceID,
			scriptPath: self.config.scriptPath,
			sensorPin: self.config.sensorPin,
			attemptNum,
			useSudo: self.config.useSudo,
			notification: "DATA_RECEIVED"
		});
	},


	/**
	 * Override the socketNotificationReceived function to handle the notifications sent from the node helper
	 *
	 * @param notification (string) The type of notification sent
	 * @param payload (any) The data sent with the notification
	 */
	socketNotificationReceived: function(notification, payload) {
		var self = this;

		// If there is no module ID sent with the notification
		if (!axis.isString(payload.original.instanceID)) {
			if (notification === "LOG") {
				if (payload.translate) { self.log(self.translate(payload.message, payload.translateVars), payload.logType); }
				else { self.log(payload.message, payload.logType); }
			}
			return;
		}

		// Filter out notifications for other instances
		if (payload.original.instanceID !== self.instanceID) {
			self.log(("Notification ignored for ID \"" + payload.original.instanceID + "\"."), "dev");
			return;
		}

		if (notification === "LOG") {
			if (payload.translate) { self.log(self.translate(payload.message, payload.translateVars), payload.logType); }
			else { self.log(payload.message, payload.logType); }
		} else if (notification === "DATA_RECEIVED") {
			if (payload.isSuccessful) {
				self.lastUpdateTime = new Date();
				self.log(self.translate("DATA_SUCCESS", { "numberOfAttempts": payload.original.attemptNum }));
				self.log(("Sensor Data: " + JSON.stringify(payload.data)), "dev");
				self.sensorData = payload.data;
				self.sensorData[self.tempUnit] += self.config.temperatureOffset;
				self.sensorData.humidity += self.config.humidityOffset;
				self.sendDataNotifications();
				self.loaded = true;
				if (self.data.position) { self.updateDom(self.config.animationSpeed); }
			} else if (payload.original.attemptNum < self.maxDataAttempts) {
				payload.error.stderr = payload.data;
				self.log(self.translate("DATA_FAILURE_RETRY", { "retryTimeInSeconds": (self.config.retryDelay / 1000) }) + "\n" + JSON.stringify(payload.error), "warn");
				setTimeout(function() { self.getData(Number(payload.original.attemptNum) + 1); }, self.config.retryDelay);
			} else {
				self.log(self.translate("DATA_FAILURE") + "\n" + JSON.stringify(payload.error), "error");
				self.loaded = true;
				if (self.data.position) { self.updateDom(self.config.animationSpeed); }
			}
		}
	},

	/**
	 * Send a notification to all the modules with the temperature and humidity.
	 * Use the INDOOR_TEMPERATURE and INDOOR_HUMIDITY notification types.
	 */
	sendDataNotifications: function() {
		var self = this;

		if (axis.isNull(self.sensorData)) { return; }

		if (self.config.sendTemperature && axis.isNumber(self.sensorData[self.tempUnit]) && !isNaN(self.sensorData[self.tempUnit])) {
			self.sendNotification("INDOOR_TEMPERATURE", self.sensorData[self.tempUnit]);
		}

		if (self.config.sendHumidity && axis.isNumber(self.sensorData.humidity) && !isNaN(self.sensorData.humidity)) {
			self.sendNotification("INDOOR_HUMIDITY", self.sensorData.humidity);
		}

	},

	/**
	 * Override the notificationReceived function.
	 * For now, there are no actions based on system or module notifications.
	 *
	 * @param notification (string) The type of notification sent
	 * @param payload (any) The data sent with the notification
	 * @param sender (object) The module that the notification originated from
	 */
	notificationReceived: function(notification, payload, sender) {
		var self = this;

		if (sender) { // If the notification is coming from another module
			if (notification === "CURRENTWEATHER_DATA") {
				if (!self.currentweatherLoaded) {
					self.currentweatherLoaded = true;
					self.sendDataNotifications();
				}
			}
		} else if (notification === "ALL_MODULES_STARTED") {

		}
	},

	/**
	 * Override the getDom function to generate the DOM objects to be displayed for this module instance
	 */
	getDom: function() {
		var self = this;
		var dataContainer, content;
		var wrapper = document.createElement("div");

		if (self.config.showTemperature || self.config.showHumidity) {

			if (!self.loaded || !self.sensorData) {
				wrapper.classList.add("loading");
				wrapper.classList.add("small");
				wrapper.innerHTML += self.translate("LOADING");
				return wrapper;
			}

			wrapper.classList.add(self.config.fontSize);
			var temperatureDecimals = self.config.roundTemperature ? 0 : 1;
			var temperatureValue = self.roundNumber(self.sensorData[self.tempUnit], temperatureDecimals).toFixed(temperatureDecimals);
			temperatureValue = self.replaceAll(temperatureValue.toString(), ".", self.config.decimalSymbol);

			var humidityDecimals = self.config.roundHumidity ? 0 : 1;
			var humidityValue = self.roundNumber(self.sensorData.humidity, humidityDecimals).toFixed(humidityDecimals);
			humidityValue = self.replaceAll(humidityValue.toString(), ".", self.config.decimalSymbol);

			if (self.config.iconView) {
				dataContainer = document.createElement("div");
				dataContainer.classList.add("icon-view-container");
				if (self.config.showTemperature) {
					var symbol = "&deg;C";
					if (self.tempUnit === "fahrenheit") { symbol = "&deg;F"; }
					else if (self.tempUnit === "kelvin") { symbol = " K"; }
					dataContainer.innerHTML += "<span class=\"fa fa-thermometer-half\"></span> " + temperatureValue + symbol;
					if (self.config.showHumidity) { dataContainer.innerHTML += " "; }
				}
				if (self.config.showHumidity) {
					dataContainer.innerHTML += "<span class=\"fa fa-tint\"></span> " + humidityValue + "%";
				}
				wrapper.appendChild(dataContainer);
			} else {
				if (self.config.showTemperature) {
					dataContainer = document.createElement("div");
					dataContainer.classList.add("temperature-container");

					content = self.config.temperatureText;
					content = content.replace(/{icon-regular-([A-Za-z0-9-_]+)}/g, "<span class=\"far fa-$1\"></span>");
					content = content.replace(/{icon-solid-([A-Za-z0-9-_]+)}/g, "<span class=\"fas fa-$1\"></span>");
					content = content.replace(/{icon-brand-([A-Za-z0-9-_]+)}/g, "<span class=\"fab fa-$1\"></span>");
					content = content.replace(/{icon-([A-Za-z0-9-_]+)}/g, "<span class=\"fa fa-$1\"></span>");
					content = self.replaceAll(content, "{temperature}", temperatureValue);
					content = self.replaceAll(content, "{humidity}", humidityValue);
					dataContainer.innerHTML = content;

					wrapper.appendChild(dataContainer);
				}

				if (self.config.showHumidity) {
					dataContainer = document.createElement("div");
					dataContainer.classList.add("humidity-container");

					content = self.config.humidityText;
					content = content.replace(/{icon-regular-([A-Za-z0-9-_]+)}/g, "<span class=\"far fa-$1\"></span>");
					content = content.replace(/{icon-solid-([A-Za-z0-9-_]+)}/g, "<span class=\"fas fa-$1\"></span>");
					content = content.replace(/{icon-brand-([A-Za-z0-9-_]+)}/g, "<span class=\"fab fa-$1\"></span>");
					content = content.replace(/{icon-([A-Za-z0-9-_]+)}/g, "<span class=\"fa fa-$1\"></span>");
					content = self.replaceAll(content, "{temperature}", temperatureValue);
					content = self.replaceAll(content, "{humidity}", humidityValue);
					dataContainer.innerHTML = content;

					wrapper.appendChild(dataContainer);
				}
			}

		} else {
			wrapper.style.display = "none";
		}

		return wrapper;
	},

	/**
	 * The roundNumber function rounds a number to the specified number of decimal places.
	 * Use a negative precision value to round to a position left of the decimal.
	 * This function overcomes the floating-point rounding issues and rounds away from 0.
	 *
	 * @param number (number) The number to round
	 * @param precision (number) The position to round to before or after the decimal
	 * @return (number) The rounded number
	 */
	roundNumber: function(number, precision) {
		if (precision >= 0) { return Number(Math.round(number + "e" + precision) + "e-" + precision); }
		else { return Number(Math.round(number + "e-" + Math.abs(precision)) + "e" + Math.abs(precision)); }
	},

	/**
	 * The replaceAll function replaces all occurrences of a string within the given string.
	 *
	 * @param str (string) The string to search within
	 * @param find (string) The string to find within str
	 * @param replace (string) The string to use as a replacement for the find string
	 * @return (string) A copy of str with all the find occurrences replaced with replace
	 */
	replaceAll: function(str, find, replace) {
		var output = "";
		var idx = str.indexOf(find);
		while (idx >= 0) {
			output += str.substr(0, idx) + replace;
			str = str.substring(idx + find.length);
			idx = str.indexOf(find);
		}
		output += str;
		return output;
	},

	/**
	 * Override the getScripts function to load additional scripts used by this module.
	 */
	getScripts: function() {
		var scripts = [];
		if (typeof axis !== "object") { scripts.push(this.file("scripts/axis.js")); }
		return scripts;
	},


	/**
	 * Override the getStyles function to load CSS files used by this module.
	 */
	getStyles: function () {
		return [
			"MMM-LocalTemperature.css",
			"font-awesome.css"
		];
	},


	/**
	 * Override the getTranslations function to load translation files specific to this module.
	 */
	getTranslations: function() {
		return {
			en: "translations/en.json"
		};
	},

	/**
	 * The log function is a convenience alias that sends a message to the console.
	 * This is an alias for the MagicMirror Log functions with a developer mode feature added.
	 * This function prepends the module name to the message.
	 *
	 * @param message (string) The message to be sent to the console
	 * @param type (string) The type of message (dev, error, info, log)
	 */
	log: function(message, type) {
		var self = this;
		if (self.config.developerMode) {
			var date = new Date();
			var time = date.getHours() + ":" + date.getMinutes() + ":" + date.getSeconds();
			message = self.name + ": (" + self.data.index + ")(" + time + ") " + message;
		} else { message = self.name + ": " + message; }
		switch (type) {
			case "error": Log.error(message); break;
			case "warn": Log.warn(message); break;
			case "info": Log.info(message); break;
			case "dev": if (self.config.developerMode) { Log.log(message); } break;
			default: Log.log(message);
		}
	}

});
